### PR TITLE
Adding k-d-m support for CIS

### PR DESCRIPTION
**Pull request by leodotcloud**
- handle info from k-d-m
- handle invalid overrideBenchmarkVersion from API

https://github.com/rancher/rancher/issues/24935
https://github.com/rancher/rancher/issues/24422

Dependencies:
- ~https://github.com/rancher/types/pull/1068~
- ~https://github.com/rancher/kontainer-driver-metadata/pull/108~

Blocked by: https://github.com/rancher/kontainer-driver-metadata/pull/124

### CODE REVIEW

File: pkg/api/customization/cluster/actions_cis.go [hunk 1]
```diff
@@ -62,11 +64,35 @@ func (a ActionHandler) runCisScan(actionName string, action *types.Action, apiCo
...
+		return httperror.WrapAPIError(err, httperror.InvalidAction,
+			fmt.Sprintf("cannot run scan on a windows cluster"))
+	}
 	if _, ok := cluster.Annotations[v3.RunCisScanAnnotation]; ok {
 		return httperror.WrapAPIError(err, httperror.Conflict,
 			fmt.Sprintf("CIS scan already running on cluster"))
 	}
 
+	if cisScanConfig.OverrideBenchmarkVersion != "" {
+		versions, err := a.CisBenchmarkVersionLister.List(namespace.GlobalNamespace, labels.Everything())
```
Comment (kinarashah) [inline pkg/api/customization/cluster/actions_cis.go]: we could directly search for OverrideBenchmarkVersion as name here?

Comment (leodotcloud) [reply pkg/api/customization/cluster/actions_cis.go]: 👍

File: pkg/controllers/user/cis/clusterScanHandler.go [hunk 2]
```diff
@@ -268,7 +293,7 @@ func (csh *cisScanHandler) deployApp(appInfo *appInfo) error {
 		"debugWorker":              appInfo.debugWorker,
 		"overrideBenchmarkVersion": appInfo.overrideBenchmarkVersion,
 	}
-	logrus.Debugf("appAnswers: %v", appAnswers)
+	logrus.Debugf("cisScanHandler: deployApp: appAnswers: %v", appAnswers)
```
Comment (kinarashah) [inline pkg/controllers/user/cis/clusterScanHandler.go]: q: is it ok to log answers object here? any data that shouldn't go in logs?

Comment (leodotcloud) [reply pkg/controllers/user/cis/clusterScanHandler.go]: there isn't anything sensitive in the answers

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 3]
```diff
@@ -100,10 +110,11 @@ func (m *MetadataController) sync(key string, setting *v3.Setting) (runtime.Obje
 	if setting == nil || (setting.Name != rkeMetadataConfig) {
 		return nil, nil
 	}
+	logrus.Infof("MetadataController: sync")
```
[the remark this answers is not in the document]
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: 👍

File: pkg/controllers/management/kontainerdrivermetadata/data.go [hunk 4]
```diff
@@ -184,6 +196,28 @@ func (md *MetadataController) saveAddons(K8sVersionedTemplates map[string]map[st
 	return nil
 }
 
+func (md *MetadataController) saveCisConfigParams(cisConfigParams map[string]v3.CisConfigParams) error {
+	for k8sVersion, cisConfigParam := range cisConfigParams {
+		logrus.Debugf("saveCisConfigParams k8sversion: %v cisConfigParam: %+v", k8sVersion, cisConfigParam)
+		if err := md.createOrUpdateCisConfigCRD(k8sVersion, cisConfigParam); err != nil {
+			logrus.Errorf("error saving cisConfig: %v", err)
```
[the remark this answers is not in the document]
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: 👍

File: pkg/controllers/management/kontainerdrivermetadata/data.go [hunk 5]
```diff
@@ -328,6 +362,88 @@ func (md *MetadataController) createOrUpdateAddonCRD(addonName, template string,
...
+			Params: cisConfigParams,
+		}
+		logrus.Debugf("driverMetadata: creating cisConfig CRD for k8sVersion: %v",
+			k8sVersion)
+		if _, err := md.CisConfig.Create(cisConfig); err != nil && !errors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+	if reflect.DeepEqual(cisConfig.Params, cisConfigParams) {
+		logrus.Debugf("driverMetadata: default cisConfig hasn't changed for k8sVersion: %v",
```
[the remark this answers is not in the document]
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: 👍

File: pkg/controllers/management/kontainerdrivermetadata/data.go [hunk 6]
```diff
@@ -328,6 +362,88 @@ func (md *MetadataController) createOrUpdateAddonCRD(addonName, template string,
...
+				Name:      k8sVersion,
+				Namespace: namespace.GlobalNamespace,
+			},
+			Params: cisConfigParams,
+		}
+		logrus.Debugf("driverMetadata: creating cisConfig CRD for k8sVersion: %v",
+			k8sVersion)
+		if _, err := md.CisConfig.Create(cisConfig); err != nil && !errors.IsAlreadyExists(err) {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: 👍

**Comment by prachidamle** [comment]
@leodotcloud can we add a test under integration for this, example a) calling API with invalid benchmark version b) API call to override benchmark version

File: pkg/controllers/management/kontainerdrivermetadata/data.go [hunk 7]
```diff
@@ -328,6 +360,87 @@ func (md *MetadataController) createOrUpdateAddonCRD(addonName, template string,
...
+			Info: cisBenchmarkVersionInfo,
+		}
+		logrus.Debugf("driverMetadata: creating cisBenchmarkVerion CRD for benchmarkVersion: %v",
+			benchmarkVersion)
+		if _, err := md.CisBenchmarkVersion.Create(cisBenchmarkVersion); err != nil && !errors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+	if reflect.DeepEqual(cisBenchmarkVersion.Info, cisBenchmarkVersionInfo) {
+		logrus.Debugf("driverMetadata: default cisBenchmarkVerionInfo hasn't changed for benchmarkVersion: %v",
```
Comment (prachidamle) [inline pkg/controllers/management/kontainerdrivermetadata/data.go]: Do we need to log this case? We need to log what has changed, but this is not required I guess

Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: ok, will remove it

**Comment by leodotcloud** [comment]
> can we add a test under integration for this, example a) calling API with invalid benchmark version b) API call to override benchmark version

When running integration tests, the cluster is not active, hence before even we can test this functionality, the code will bail out. Hence this test needs to be added in validation.

File: pkg/controllers/management/kontainerdrivermetadata/data.go [hunk 8]
```diff
@@ -328,6 +360,85 @@ func (md *MetadataController) createOrUpdateAddonCRD(addonName, template string,
...
+				Name:      benchmarkVersion,
+				Namespace: namespace.GlobalNamespace,
+			},
+			Info: cisBenchmarkVersionInfo,
+		}
+		logrus.Debugf("driverMetadata: creating cisBenchmarkVerion CRD for benchmarkVersion: %v",
+			benchmarkVersion)
+		if _, err := md.CisBenchmarkVersion.Create(cisBenchmarkVersion); err != nil && !errors.IsAlreadyExists(err) {
+			return err
+		}
```
Comment (dramich) [inline pkg/controllers/management/kontainerdrivermetadata/data.go]: This needs a return nil

File: pkg/controllers/management/kontainerdrivermetadata/data.go [hunk 9]
```diff
@@ -328,6 +360,85 @@ func (md *MetadataController) createOrUpdateAddonCRD(addonName, template string,
 	return nil
 }
 
+func (md *MetadataController) createOrUpdateCisConfigCRD(
```
Comment (dramich) [inline pkg/controllers/management/kontainerdrivermetadata/data.go]: nit: CRD should not be in the name of the func

File: pkg/controllers/management/kontainerdrivermetadata/data.go [hunk 10]
```diff
@@ -328,6 +360,85 @@ func (md *MetadataController) createOrUpdateAddonCRD(addonName, template string,
...
+		k8sVersion, cisConfig.Params, cisConfigParams)
+	cisConfigCopy := cisConfig.DeepCopy()
+	cisConfigCopy.Params = cisConfigParams
+	if _, err := md.CisConfig.Update(cisConfigCopy); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (md *MetadataController) createOrUpdateCisBenchmarkVersionCRD(
```
Comment (dramich) [inline pkg/controllers/management/kontainerdrivermetadata/data.go]: nit: CRD should not be in the name of the func

File: pkg/controllers/management/kontainerdrivermetadata/data.go (hunk 8 above)
Comment (kinarashah) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: heh didn't we already fix this @leodotcloud 😆

File: pkg/controllers/management/kontainerdrivermetadata/data_getter.go [hunk 11]
```diff
@@ -16,6 +16,25 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+func GetCisConfigParams(
```
Comment (dramich) [inline pkg/controllers/management/kontainerdrivermetadata/data_getter.go]: Why not just hang this off of `cisScanHandler` and keep it in the same package?

File: pkg/controllers/management/kontainerdrivermetadata/data.go (hunk 8 above)
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: crap, I missed the other CRD.

Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: will fix it. thanks @dramich

File: pkg/controllers/management/kontainerdrivermetadata/data_getter.go (hunk 11 above)
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data_getter.go]: Following the same pattern as the rest of the file.

File: pkg/controllers/management/kontainerdrivermetadata/data.go (hunk 9 above)
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: I think it's easier to read and understand the code with CRD in place, let's leave it this way :)

File: pkg/controllers/management/kontainerdrivermetadata/data.go (hunk 10 above)
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: I think it's easier to read and understand the code with CRD in place, let's leave it this way :)

**Comment by dramich** [comment]
LGTM after the return is added.

File: pkg/controllers/management/kontainerdrivermetadata/data.go (hunk 8 above)
Comment (leodotcloud) [reply pkg/controllers/management/kontainerdrivermetadata/data.go]: done added

**Pull request merged**